### PR TITLE
Refactored input handling and added key mapping for road laying

### PR DESCRIPTION
**Pull request by kingoflolz**
Should be *slightly* less of a hack than before

### CODE REVIEW

**Comment by kingoflolz** [comment]
Not happy about some code duplication, is it possible to send a Vec in a message?

**Comment by aeickhoff** [comment]
Just use a `CompactVec` (I always import it as `CVec`) and have your message auto-derive `Compact, Clone`

File: src/core/ui.rs [hunk 1]
```diff
@@ -328,24 +288,27 @@ impl Recipient<UIUpdate> for UserInterface {
...
-                    if self.input_state.yaw_mod || self.input_state.pitch_mod || self.input_state.pan_mod {
-                        if self.input_state.yaw_mod {
+                    if yaw_mod || pitch_mod || pan_mod {
+                        if yaw_mod {
                             Renderer::id() << MoveEye { scene_id: 0, movement: ::monet::Movement::Yaw(
                                 -delta.x * self.settings.rotation_speed * inverted/ 300.0)
                             };
                             
                             self.cursor_2d = position;
                         }
```
Comment (Herbstein) [inline src/core/ui.rs]: I think the *_mod if-statements are doing early returns? Shouldn't they be terminated with a semi-colon?

Comment (kingoflolz) [inline src/core/ui.rs]: line 295? I don't think you can put semicolons in if statements?

Comment (Herbstein) [inline src/core/ui.rs]: I was thinking at the end, so 302 in the above snippet.

Comment (Herbstein) [inline src/core/ui.rs]: Nevermind. I though I remembered having that exact bug when changing the movement code myself. You're all good!

Comment (kingoflolz) [inline src/core/ui.rs]: Hrm, it works, so it's not returning early. but I will add semicolons to clarify my intent. Thanks for pointing this out!

**Comment by aeickhoff** [comment]
Looks good so far, but it would be really cool if LaneStrokeCanvas could not care about settings and wouldn't have to store key state itself, but instead just got something KeyUp/Down messages already with "semantic meaning" (like CreateGridButtonDown) from the UI.

Is it clear what I mean and how would you architect that?

**Comment by kingoflolz** [comment]
Hrm, then instead of the LaneStrokeCanvas caring about settings, UI would have to care about LaneStrokeCanvas.

If we were to do that, we can just decode keystrokes into things with meaning. We can change Event3d to have semantic meaning, maybe to LaneStrokeCanvas_UI_events or something, and have a different struct for every single type of interactable (or a single gigantic enum).

**Comment by aeickhoff** [comment]
You're right, keep the design like it is now. But then it would be nice if Settings would allow "registering" semantic buttons and default bindings, so it would not have to know about all intractables.

**Comment by kingoflolz** [comment]
I think that is a good idea, but to do that we will need to refactor the camera into an Interactable.

P.S. What if an interactable needs text entry? Would it have to register every single key? Maybe it would be a better idea to have multiple interactable traits which receive different messages, so there would be ButtonInteractable, MouseInteractable and TextInteractable?

**Comment by aeickhoff** [comment]
Captured that in #91

**Comment by aeickhoff** [comment]
💣not using rustfmt yet 💣
Please install & apply rustfmt using:
- `cargo install rustfmt`
- `rustfmt --write-mode=overwrite ./src/main.rs ./lib/*/src/lib.rs`

then merge master and there should be less nasty conflicts

**Comment by kingoflolz** [comment]
Taking another shot at it, closing for now

**Pull request closed without merging**